supervisor/core/supervisor.go

```diff
@@ -0,0 +1,208 @@
+func (sv *Supervisor) StartInstance(name string, env []string, isRestart bool) (int, error) {
+	// When Supervisor is terminating, we will lock "termMutex"
+	// to prevent new instances from starting during Supervisor termination.
+	sv.termMutex.RLock()
+	defer sv.termMutex.RUnlock()
+
+	// Form the path to the instance and check is it exists.
+	instPath := path.Join(sv.cfg.InstancesDir, name+".lua")
+	if _, err := exec.LookPath(instPath); err != nil {
+		return 0, err
+	}
+
+	// Start an Instance.
+	cmd := exec.Command(instPath)
+	cmd.Env = append(os.Environ(), env...)
+	inst := NewInstance(name, cmd, env, isRestart)
+	if err := inst.Start(); err != nil {
+		return 0, err
+	}
+
+	return sv.addInstance(inst), nil
+}
+
+// RestartAfterTermInstance should be used to restart an instance in case
+// of an unexpected termination of the instance.
+// First of all, the method checks if the instance is alive. In this case,
+// the method returns the current Instance ID with an error.
+// If the Instance has been terminated and Instance.IsRestart is true,
+// then the Instance will be restarted.
+// On success returns the Instance ID (0 invalid).
+func (sv *Supervisor) RestartAfterTermInstance(pid int) (int, error) {
+	// Find an Instance.
+	id, inst := sv.getInstanceByPid(pid)
+	if inst == nil {
+		return 0, errors.New("Unknown Instance.")
+	}
+
+	// We don't want to restart the Instance at the same time
+	// as StopAllInstance is running.
+	sv.termMutex.RLock()
+	defer sv.termMutex.RUnlock()
+
+	// If the Instance is alive something went wrong.
+	if inst.IsAlive() {
+		return id, errors.New("Instance is alive.")
+	}
+
+	if !inst.IsRestart {
+		return id, errors.New("The IsRestart Instance flag is false.")
+	}
+
+	if err := inst.Restart(); err != nil {
+		return id, err
+	}
+
+	return id, nil
+}
+
+// StopInstance terminate an Instance by ID.
+func (sv *Supervisor) StopInstance(id int, force bool) error {
+	// When Supervisor is terminating, we will lock "termMutex"
+	// to prevent an instance termination from several places
+	// (see StopAllInstance).
+	sv.termMutex.RLock()
+	defer sv.termMutex.RUnlock()
+
+	inst := sv.getInstance(id)
+	if inst == nil {
+		return errors.New("Unknown instance with id " + strconv.Itoa(id))
+	}
+
+	if err := inst.Stop(sv.cfg.TermTimeout, force); err != nil {
+		return err
+	}
+	sv.deleteInstance(id)
+
+	return nil
+}
+
+// StopAllInstance terminate all Instances managed by the Supervisor.
+func (sv *Supervisor) StopAllInstance() {
+	// Disable start / stop Instances.
+	sv.termMutex.Lock()
+
+	// Start termination of all Instances.
+	var wg sync.WaitGroup
+	// We don't want to delete instances from the map
+	// until the iteration is done.
+	sv.instMutex.RLock()
+	for id, inst := range sv.instancesById {
+		wg.Add(1)
+		go func(id int, inst *Instance) {
+			inst.Stop(sv.cfg.TermTimeout, true)
+			sv.deleteInstance(id)
+			wg.Done()
+		}(id, inst)
+	}
+	sv.instMutex.RUnlock()
+	defer sv.termMutex.Unlock()
```

Comment:
This `defer` should be moved to appropriate lock.

BTW, why don't you use `defer` for sv.instMutex.RUnlock()? Seems that it should be unlocked after `wg.Wait()`.

Reply:
You need to Unlock the `sv.instMutex` mutex before `wg.Wait()`, because while it is Lock, no one goroutines created in the `for` loop will be done (see the "deleteInstance" method).

Reply:
So, you use `sv.instMutex` in two places just to delay instances deletion until all goroutines are created?
You've locked `termMutex`, so starting/stopping instances is disabled and all this commands wouldn't modify instances map. `instMutex` is locked in `deleteInstance` , and it;s Rlocked by List and Status. Both this commands can show intermediate results while `Stop all instances` is running.

Reply:
> So, you use `sv.instMutex` in two places just to delay instances deletion until all goroutines are created?

If I understood you correctly, yep. While I run the goroutines, I iterate over the `instances` map and want that this map not change.

> You've locked `termMutex`, so starting/stopping instances is disabled and all this commands wouldn't modify instances map. `instMutex` is locked in `deleteInstance` , and it;s Rlocked by List and Status. Both this commands can show intermediate results while `Stop all instances` is running.

Yep and it's ok. This will be honest list and status at the time then it was requested.

Reply:
> While I run the goroutines, I iterate over the instances map and want that this map not change.
> You've locked termMutex, so starting/stopping instances is disabled and all this commands wouldn't modify instances map. 

Seems that this lock can be removed.

Reply:
The map will modify from goroutines (`deleteInstance`).